libdrgn/program.c

```diff
@@ -742,34 +823,281 @@ static struct drgn_error *drgn_program_cache_prstatus(struct drgn_program *prog)
+
+	if ((prog->flags &
+	     (DRGN_PROGRAM_IS_LINUX_KERNEL | DRGN_PROGRAM_IS_LIVE)) ==
+	    DRGN_PROGRAM_IS_LIVE) {
+		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+					 "finding threads is not yet supported for live processes");
+	}
+
+	*ret = malloc(sizeof(**ret));
+	if (!*ret)
+		return &drgn_enomem;
+	(*ret)->prog = prog;
+	if (prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+		err = linux_helper_task_iterator_init(&(*ret)->task_iter, prog);
+		if (err)
+			goto err;
+		drgn_object_init(&(*ret)->entry.object, prog);
+		(*ret)->entry.prstatus = (struct nstring){};
 	} else {
-		prog->prstatus_cached = true;
+		err = drgn_program_cache_core_dump_notes(prog);
+		if (err)
+			goto err;
+		(*ret)->iterator = drgn_thread_set_first(&prog->thread_set);
 	}
+	return NULL;
+
+err:
+	free(*ret);
 	return err;
 }
 
+LIBDRGN_PUBLIC void
+drgn_thread_iterator_destroy(struct drgn_thread_iterator *it)
+{
+	if (it) {
+		if (it->prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+			drgn_object_deinit(&it->entry.object);
+			linux_helper_task_iterator_deinit(&it->task_iter);
+		}
+		free(it);
+	}
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_thread_iterator_next(struct drgn_thread_iterator *it,
+			  struct drgn_thread **ret)
+{
+	if (it->prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+		const struct drgn_object *task;
+		struct drgn_error *err =
+			linux_helper_task_iterator_next(&it->task_iter, &task);
+		if (err)
+			return err;
+		if (!task) {
+			*ret = NULL;
+			return NULL;
+		}
+		it->entry.prog = drgn_object_program(task);
+		err = drgn_object_copy(&it->entry.object, task);
+		if (err)
+			return err;
+		struct drgn_object tid;
+		union drgn_value tid_value;
+		drgn_object_init(&tid, drgn_object_program(task));
+		err = drgn_object_member_dereference(&tid, task, "pid");
+		if (!err)
+			err = drgn_object_read_integer(&tid, &tid_value);
+		drgn_object_deinit(&tid);
+		if (err)
+			return err;
+		it->entry.tid = tid_value.uvalue;
+		*ret = &it->entry;
+		return NULL;
+	} else {
+		*ret = it->iterator.entry;
+		if (it->iterator.entry)
+			it->iterator = drgn_thread_set_next(it->iterator);
+	}
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_thread(struct drgn_program *prog, uint32_t tid,
+			 struct drgn_thread **ret)
+{
+	struct drgn_error *err;
+
+	if (prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) {
+		*ret = malloc(sizeof(**ret));
+		if (!*ret)
+			return &drgn_enomem;
+		(*ret)->prog = prog;
+		(*ret)->tid = tid;
+		(*ret)->prstatus = (struct nstring){};
+
+		struct drgn_object *object = &(*ret)->object;
+		drgn_object_init(object, prog);
+		err = drgn_program_find_object(prog, "init_pid_ns", NULL,
+					       DRGN_FIND_OBJECT_VARIABLE,
+					       object);
+		if (err)
+			goto kernel_err;
+		err = drgn_object_address_of(object, object);
+		if (err)
+			goto kernel_err;
+		err = linux_helper_find_task(object, object, tid);
+		if (err)
+			goto kernel_err;
+		return NULL;
+
+kernel_err:
+		drgn_object_deinit(object);
```

Comment:
Could just call `drgn_thread_destroy(*ret)` instead of manually reproducing its contents here.